### PR TITLE
Add support for tunnel route config that has both match group and port variable

### PR DESCRIPTION
**Pull request by lzx404243**
This PR adds support for the `tunnel_route` configuration that contains both the regex match group number and the port variable specification, such as the following
```
- fqdn: "node.*.example.com"
  tunnel_route: "backend.$1.example.com:{proxy_protocol_port}"
```
Prior to this PR, only one of the match group specification or port variable specification can be used in the `tunnel_route` configuration.

### CODE REVIEW

**Review by bneradt — CHANGES_REQUESTED** [review]
Thanks for fixing this @lzx404243.

File: iocore/net/P_SNIActionPerformer.h [hunk 1]
```diff
@@ -232,8 +241,11 @@ class TunnelDestination : public ActionItem
   YamlSNIConfig::TunnelPreWarm tunnel_prewarm = YamlSNIConfig::TunnelPreWarm::UNSET;
   const std::vector<int> &alpn_ids;
 
-  OpId fnArrIndex{OpId::DEFAULT}; /// On creation, we decide which function needs to be called, set the index and then we
-                                  /// call it with the relevant data
+  // The indexes of the mapping functions that need to be called. On
+  // creation, we decide which functions need to be called, add the
+  // coressponding indexes and then we call those functions with the relevant
+  // data.
+  std::vector<OpId> fnArrIndexes;
```
[the remark this answers is not in the document]
Comment (lzx404243) [reply iocore/net/P_SNIActionPerformer.h]: Good point. Fixed

**Review by bneradt — APPROVED** [review]

File: doc/admin-guide/files/sni.yaml.en.rst [hunk 2]
```diff
@@ -176,6 +176,8 @@ tunnel_route              Inbound   Destination as an FQDN and port, separated b
                                     Protocol <proxy-protocol>` for more information on Proxy Protocol and how it is
                                     configured for |TS|.
 
+                                    Note that the match group number can be used in combination with the ``{inbound_local_port}`` and ``{proxy_protocol_port}`` literal strings.
+
```
Comment (cmcfarlen) [inline doc/admin-guide/files/sni.yaml.en.rst:180]: The code and this comment don't seem to agree.  The code is such that only one of `{inbound_local_port}` or `{proxy_protocol_port}` will be used with `{inbound_local_port}` taking precedence.  If both are specified, the `{proxy_protocol_port}` function is not run at all.  I recommend address this in the documentation and better handle the case where both strings are present (warning or error).

**Review by cmcfarlen — CHANGES_REQUESTED** [review]
The documentation could better match the code and I feel the code should handle the case where both port strings appear (even if that means a warning or error).

File: doc/admin-guide/files/sni.yaml.en.rst (hunk 2 above)
Comment (brbzull0) [reply doc/admin-guide/files/sni.yaml.en.rst:180]: yes, if both are present I also suggest adding an error as that seems not allowed. It should be either one `proxy_protocol_port` or `inbound_local_port` with a combination of the groups.

**Comment by SolidWallOfCode** [comment]
I'll defer to @brbzull0 and @cmcfarlen but otherwise I think it's fine.

File: doc/admin-guide/files/sni.yaml.en.rst (hunk 2 above)
Comment (lzx404243) [reply doc/admin-guide/files/sni.yaml.en.rst:180]: Thanks for reviewing this and  good point on handling the case where both port variables are specified. I updated the doc to make it more explicit that only one can be specified. I also added a check in the code that would emit an error message when both are present.

File: iocore/net/P_SNIActionPerformer.h [hunk 3]
```diff
@@ -108,14 +109,20 @@ class TunnelDestination : public ActionItem
...
+    auto pp_port_start_pos   = destination.find(MAP_WITH_PROXY_PROTOCOL_PORT_STR);
+    bool has_recv_port_var   = recv_port_start_pos != std::string::npos;
+    bool has_pp_port_var     = pp_port_start_pos != std::string::npos;
+    if (has_recv_port_var && has_pp_port_var) {
+      Error("Invalid destination \"%.*s\" in SNI configuration - Only one port variable can be specified.",
+            static_cast<int>(destination.size()), destination.data());
+    } else if (has_recv_port_var || has_pp_port_var) {
+      // Specified only one of the port variables.
+      fnArrIndexes.push_back(has_recv_port_var ? OpId::MAP_WITH_RECV_PORT : OpId::MAP_WITH_PROXY_PROTOCOL_PORT);
+      var_start_pos = has_recv_port_var ? recv_port_start_pos : pp_port_start_pos;
```
Comment (cmcfarlen) [inline iocore/net/P_SNIActionPerformer.h:124]: Its more readable with separate conditions like:

```
if (recv && pp) {
 //error
} else if (recv) {
} else if (pp) {
}
```

**Review by cmcfarlen — DISMISSED** [review]
The new code to add the function is a little hard to read imo.

File: iocore/net/P_SNIActionPerformer.h (hunk 3 above)
Comment (lzx404243) [reply iocore/net/P_SNIActionPerformer.h]: Yeah that looks better. Updated.

**Pull request merged**